packages/toolpad-app/src/toolpad/Home/index.tsx

```diff
@@ -125,37 +135,67 @@ function CreateAppDialog({ onClose, ...props }: CreateAppDialogProps) {
 
   const isFormValid = Boolean(name);
 
-  return (
-    <Dialog {...props} onClose={config.isDemo ? NO_OP : onClose} maxWidth="xs">
-      <DialogForm
-        onSubmit={async (event) => {
-          invariant(isFormValid, 'Invalid form should not be submitted when submit is disabled');
-
-          event.preventDefault();
-          let recaptchaToken;
-          if (config.recaptchaSiteKey) {
-            await new Promise<void>((resolve) => {
-              grecaptcha.ready(resolve);
-            });
-            recaptchaToken = await grecaptcha.execute(config.recaptchaSiteKey, {
-              action: 'submit',
+  const handleSubmit = React.useCallback(
+    async (event: React.FormEvent<HTMLFormElement>) => {
+      invariant(isFormValid, 'Invalid form should not be submitted when submit is disabled');
+
+      event.preventDefault();
+
+      const latestRecaptchaWidgetId = latestRecaptchaWidgetIdRef.current;
+      const hasShownRecaptchaCheckbox = latestRecaptchaWidgetId !== null;
+
+      // Check if captcha checkbox was solved
+      let recaptchaToken = hasShownRecaptchaCheckbox
+        ? grecaptcha.getResponse(latestRecaptchaWidgetId)
+        : '';
+
+      if (config.recaptchaV3SiteKey && !hasShownRecaptchaCheckbox) {
+        // Invisible captcha validation
+        await new Promise<void>((resolve) => {
+          grecaptcha.ready(resolve);
+        });
+        recaptchaToken = await grecaptcha.execute(config.recaptchaV3SiteKey, {
+          action: 'submit',
+        });
+      }
+
+      const appDom = dom.trim() ? JSON.parse(dom) : null;
+
+      try {
+        await createAppMutation.mutateAsync([
+          name,
+          {
+            from: {
+              ...(appDom ? { kind: 'dom', dom: appDom } : { kind: 'template', id: appTemplateId }),
+            },
+            ...(recaptchaToken
+              ? {
+                  captcha: {
+                    token: recaptchaToken,
+                    version: hasShownRecaptchaCheckbox ? 2 : 3,
+                  },
+                }
+              : {}),
+          },
+        ]);
+      } catch (error) {
+        if (config.recaptchaV2SiteKey && !hasShownRecaptchaCheckbox) {
+          if (error instanceof ApiError && error.code === API_ERROR_CODES.VALIDATE_CAPTCHA_FAILED) {
+            // Show captcha checkbox
+            const widgetId = grecaptcha.render(CAPTCHA_TARGET_ID, {
```

Comment:
We can't use a ref to a dom element here?

Reply:
You mean for the id? Yeah we could, it would be a bit better so will do!